Add Prototype patters and contribute with issue #26

mijuanmontalvo opened the pull request:
To contribute with issue #26, I did this:

In the "1. Creational Patterns" section table of the README.md file in the "patterns" folder add a link to the word Prototype, so that when users click, it will take them to the "Prototype" pattern folder

Inside the "patterns" folder I created a folder called "Prototype" (the folder indicated in the previous paragraph).

Inside the "Prototype" folder I created two files "README.md" and "Prototype.js".

In the "README.md" file I have placed a definition of the prototype pattern and an explanation of an example of this type of pattern for javascrip.

In the "Prototype.js" file add the example code of the prototype pattern for javascrip

mijuanmontalvo committed:
Create Prototype folder and its README.md file

mijuanmontalvo committed:
Merge branch 'nic-dgl104-winter-2024:main' into Contribution1

mijuanmontalvo committed:
Create README file with information about Prototype pattern

mijuanmontalvo committed:
Create file (Js) with code example about prototype pattern

mijuanmontalvo committed:
Update format of README.md file

mijuanmontalvo committed:
Update format of README.md

mijuanmontalvo committed:
Update code of README.md

ash-teach commented on patterns/Prototype/Prototype.js:
```diff
@@ -0,0 +1,20 @@
+  
+  // Create new instances using the prototype
+  const car1 = Object.create(vehiclePrototype);
+  car1.init('Toyota', 'Camry');
```
One of the benefits of the prototype pattern is that repeated initialization of common objects is less cumbersome because you can reuse some initial values and only change the necessary ones (in contrast to instantiating a new class object, which typically always requires full initialization).

In your examples here, you've sidestepped this benefit, since you've initialized both objects from scratch. It may be preferable to show an example that demonstrates a more complex base object (with more parameters) and smaller changes in the copied objects.

ash-teach commented on patterns/Prototype/Prototype.js:
```diff
@@ -0,0 +1,20 @@
+  car2.init('Honda', 'Civic');
+  
+  console.log(car1.getDetails()); // Output: 'Toyota Camry'
+  console.log(car2.getDetails()); // Output: 'Honda Civic'
```
You could probably do a bit more to demonstrate that this conforms to the prototype pattern. In particular, you could compare the two car objects to demonstrate that they are distinct, but then also compare their respective `prototype` values (see more info below) to show that they are derived from the same base object:
```javascript
console.log(car1 === car2) //prints true
console.log(car1.prototype === car2.prototype) //prints false
```

ash-teach commented on patterns/Prototype/README.md:
```diff
@@ -0,0 +1,112 @@
+
+
+The Prototype Design Pattern is a creational pattern that enables the creation of new objects by copying an existing object. Prototype allows us to hide the complexity of making new instances from the client. The concept is to copy an existing object rather than create a new instance from scratch, something that may include costly operations. The existing object acts as a prototype and contains t ...
+<br>
```
Please remove all HTML tags, such as `<br>`. They are not necessary. If you want a space between paragraphs, just make sure there is an empty line between them in the markdown.

ash-teach commented on patterns/Prototype/README.md:
```diff
@@ -0,0 +1,112 @@
+## Prototype Pattern in JavaScript
+
+It promotes the creation of objects without specifying the exact class of object to create. This pattern is particularly useful when you want to create instances of complex objects efficiently.
+<br>
```
What is particularly interesting about this example is the fact that JavaScript is a prototypal language. This means that all JavaScript objects _have_ a prototype already (you can see it by using `.prototype` on any JS object), and so the prototype design pattern is supported in JavaScript out of the box.

In other words, you don't really need to do anything in JS to implement the prototype pattern - it's already there via `Object.create()` and `Object.assign()`! Contrast the JavaScript implementation to the Java implementation, for example - the Java implementation is necessarily much more complex, because Java does not have a prototype object.

However, I think this means you have another responsibility here: You should state that as a prototypal language, JS supports prototype pattern without any real additional code needed. See if you can find some sources to help you state this clearly here.

ash-teach commented on patterns/Prototype/README.md:
````diff
@@ -0,0 +1,112 @@
+```
+
+Below is an explanation of the above code.
+<br>
````
You've copy pasted the code here many more times than is necessary. I don't need to see it copies here when it's available in the .js file. Please remove the full copy paste from above this comment.

ash-teach commented on patterns/Prototype/README.md:
````diff
@@ -0,0 +1,112 @@
+  },
+};
+
+```
````
As above, I don't think you need to copy paste _all_ the code since it's easily accessible in the .js file. Instead, focus your explanation on the more important parts (i.e. the use of `Object.create`).

ash-teach commented on patterns/Prototype/README.md:
```diff
@@ -0,0 +1,112 @@
+
+- In this part of the code, new vehicle instances (**car1** and **car2**) are created using the **Object.create()** method.
+- The **vehiclePrototype** object is passed as an argument to **Object.create()**, which sets the prototype of the new created object to **vehiclePrototype**.
+- The **init()** method is then called on each instance to initialize the **make** and **model** properties with specific values.
```
When writing about code in line don't use **bold emphasis**. Instead write `in-line code`. You do this by surrounding the code in single backticks. So your third bullet above should appear as follows:

> The `init()` method is then called on each instance to initialize the `make` and `model` properties with specific values.

ash-teach commented on patterns/Prototype/README.md:
```diff
@@ -0,0 +1,112 @@
+<br>
+<br>
+
+# Resouces
```
Spelling. "Resources"

ash-teach commented on patterns/README.md:
```diff
@@ -25,7 +25,7 @@ These patterns deal with object creation mechanisms, trying to create objects in
 | Builder         | Separates the construction of a complex object from its representation, allowing the same construction process to create various representations.|
 | Factory Method  | Defines an interface for creating an object but lets subclasses alter the type of objects that will be created.     |
-| Prototype       | Creates new objects by copying an existing object, known as the prototype.  
+| [Prototype](./Prototype/)       | Creates new objects by copying an existing object, known as the prototype.  
```
I think you should move this into a different folder: patterns\Creational\Prototype

ash-teach commented:
Thanks @mijuanmontalvo - lots of great work here. I think there are some improvements to be made, especially relating to the nature of prototypes in JavaScript. I've asked for some fixes in the comments above - you may not be able to address them all at this late stage, but do your best. If you need help, feel free to request it from others. If you need help prioritizing the tasks listed in the review comments, just let me know.

mijuanmontalvo committed:
Merge branch 'nic-dgl104-winter-2024:main' into main

mijuanmontalvo committed:
Update branch to include observations from PR #76

mijuanmontalvo committed:
Update format

mijuanmontalvo committed:
Merge branch 'Contribution1'

mijuanmontalvo commented:
Hello @ash-teach, how are you, 

Please review your comments on this pull request and make the following changes:

- In the README.md file of the prototype pattern, add a section that includes a java example and add a file with a java example.
- In the README.md file I removed all HTML tags `<br>`
- I modified the content of the JavaScript prototype pattern, now it is gone, there is no code in the README.md file, all the code is in the .js file
- Regarding the inline code, I removed the bold emphasis, now the code words are in single backticks.
- I fixed the spelling mistake (Resources), now it shows "Resources"
- I moved all the prototype pattern information to the patterns\Creational\Prototype folder and set the links to point to this new path.

All the previous items were made in my "main" branch so I understand that those changes are automatically reflected in this pull request. Please review these changes, thank you very much
